File: src/main/java/org/simbrain/plot/projection/ProjectionDesktopComponent.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.plot.projection;

import org.jfree.chart.ChartFactory;
import org.jfree.chart.ChartPanel;
import org.jfree.chart.JFreeChart;
import org.jfree.chart.labels.CustomXYToolTipGenerator;
import org.jfree.chart.labels.StandardXYItemLabelGenerator;
import org.jfree.chart.labels.XYItemLabelGenerator;
import org.jfree.chart.plot.PlotOrientation;
import org.jfree.chart.renderer.xy.XYLineAndShapeRenderer;
import org.jfree.data.xy.XYDataset;
import org.jfree.data.xy.XYSeries;
import org.jfree.data.xy.XYSeriesCollection;
import org.simbrain.plot.actions.PlotActionManager;
import org.simbrain.util.ResourceManager;
import org.simbrain.util.SimbrainPreferences;
import org.simbrain.util.Utils;
import org.simbrain.util.genericframe.GenericFrame;
import org.simbrain.util.projection.*;
import org.simbrain.util.widgets.ShowHelpAction;
import org.simbrain.workspace.component_actions.CloseAction;
import org.simbrain.workspace.gui.DesktopComponent;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.FocusAdapter;
import java.awt.event.FocusEvent;
import java.awt.geom.Ellipse2D;
import java.util.Map.Entry;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Gui Component to display a high dimensional projection object.
 */
public class ProjectionDesktopComponent extends DesktopComponent<ProjectionComponent> {

    /**
     * The JFreeChart chart.
     */
    private JFreeChart chart;

    /**
     * JChart representation of the data.
     */
    private XYSeriesCollection xyCollection;

    /**
     * Executor service for running iterable projection methods.
     */
    private ExecutorService executor = Executors.newSingleThreadExecutor();

    /**
     * List of projector types.
     */
    private JComboBox<String> projectionList = new JComboBox<String>();

    /**
     * Iterate once.
     */
    protected JButton iterateBtn = new JButton(ResourceManager.getImageIcon("menu_icons/Step.png"));

    /**
     * Play button.
     */
    private JButton playBtn = new JButton(ResourceManager.getImageIcon("menu_icons/Play.png"));

    /**
     * Points indicator.
     */
    private JLabel pointsLabel = new JLabel();

    /**
     * Dimension indicator.
     */
    private JLabel dimsLabel = new JLabel(" Dimensions:");

    /**
     * Error indicator.
     */
    private JLabel errorLabel = new JLabel();

    /**
     * Error bar.
     */
    private JToolBar errorBar = new JToolBar();

    /**
     * Show error option.
     */
    private boolean showError = true;

    /**
     * Warning label.
     */
    private JLabel warningLabel = new JLabel(ResourceManager.getImageIcon("menu_icons/Warning.png"));

    /**
     * Combo box for first dimension of coordinate projection.
     */
    private JComboBox<Integer> adjustDimension1 = new JComboBox<>();

    /**
     * Model for adjustDimension1.
     */
    private DefaultComboBoxModel<Integer> adjustDimension1Model = new DefaultComboBoxModel<Integer>();

    /**
     * Combo box for first dimension of coordinate projection.
     */
    private JComboBox<Integer> adjustDimension2 = new JComboBox<Integer>();

    /**
     * Model for adjustDimension2.
     */
    private DefaultComboBoxModel<Integer> adjustDimension2Model = new DefaultComboBoxModel<Integer>();

    /**
     * Panel for showing Sammon step size and label, both with tooltip.
     */
    private Box sammonStepSizePanel = Box.createHorizontalBox();

    /**
     * Construct the Projection GUI.
     */
    public ProjectionDesktopComponent(final GenericFrame frame, final ProjectionComponent component) {
        super(frame, component);
        setPreferredSize(new Dimension(500, 400));
        setLayout(new BorderLayout());

        Projector proj = getWorkspaceComponent().getProjector();

        // Generate the graph
        xyCollection = new XYSeriesCollection();
        xyCollection.addSeries(new XYSeries("Data", false, true));
        chart = ChartFactory.createScatterPlot("", "Projection X", "Projection Y", xyCollection, PlotOrientation.VERTICAL, false, true, false);
        // chart.getXYPlot().getDomainAxis().setRange(-100, 100);
        // chart.getXYPlot().getRangeAxis().setRange(-100, 100);
        chart.getXYPlot().setBackgroundPaint(Color.white);
        chart.getXYPlot().setDomainGridlinePaint(Color.gray);
        chart.getXYPlot().setRangeGridlinePaint(Color.gray);
        chart.getXYPlot().getDomainAxis().setAutoRange(true);
        chart.getXYPlot().getRangeAxis().setAutoRange(true);
        chart.getXYPlot().setForegroundAlpha(.5f); // TODO: Make this settable

        ChartPanel chartPanel = new ChartPanel(chart);

        // Custom render points as dots (not squares) and use custom tooltips
        // that show high-d point
        CustomRenderer renderer = new CustomRenderer();
        chart.getXYPlot().setRenderer(renderer);
        // TODO: Make the visibility of series lines and point size adjustible
        renderer.setSeriesLinesVisible(0, false);
        renderer.setSeriesShape(0, new Ellipse2D.Double(-7, -7, 7, 7));
        CustomToolTipGenerator generator = new CustomToolTipGenerator();
        renderer.setSeriesToolTipGenerator(0, generator);

        // TODO
        // Set up custom label generator, used with some couplings
        // renderer.setBaseItemLabelsVisible(true);
        // renderer.setBaseItemLabelGenerator(new LegendXYItemLabelGenerator());

        // Toolbar
        JButton openBtn = new JButton(ResourceManager.getImageIcon("menu_icons/Open.png"));
        openBtn.setToolTipText("Open high-dimensional data");
        JButton saveBtn = new JButton(ResourceManager.getImageIcon("menu_icons/Save.png"));
        saveBtn.setToolTipText("Save data");
        projectionList.setMaximumSize(new java.awt.Dimension(200, 100));

        iterateBtn = new JButton(ResourceManager.getImageIcon("menu_icons/Step.png"));
        iterateBtn.addActionListener(e -> {
            proj.iterate();
            update();
        });

        playBtn = new JButton(ResourceManager.getImageIcon("menu_icons/Play.png"));
        playBtn.addActionListener(e -> {
            if (proj.isRunning()) {
                // If already running, image is stop. Click and it should stop the algorithm and
                // change the image to play.
                playBtn.setToolTipText("Start iterating projection algorithm");
                playBtn.setIcon(ResourceManager.getImageIcon("menu_icons/Play.png"));
                proj.setRunning(false);
            } else {
                // If not running, image is play. Click and it should run the algorithm
                // and change the image to stop.
                playBtn.setIcon(ResourceManager.getImageIcon("menu_icons/Stop.png"));
                playBtn.setToolTipText("Stop iterating projection algorithm");
                proj.setRunning(true);
                executor.execute(() -> {
                    while (proj.isRunning()) {
                        proj.iterate();
                        proj.getEvents().fireDataChanged();
                    }
                });
            }
        });

        JButton clearBtn = new JButton(ResourceManager.getImageIcon("menu_icons/Eraser.png"));
        clearBtn.addActionListener(e ->
                SwingUtilities.invokeLater(() -> {
                    getWorkspaceComponent().clearData();
                    xyCollection.getSeries(0).clear();
                }));

        JButton prefsBtn = new JButton(ResourceManager.getImageIcon("menu_icons/Prefs.gif"));
        prefsBtn.addActionListener(e -> {
            // TODO (Still working out overall dialog structure).
        });

        JButton randomBtn = new JButton(ResourceManager.getImageIcon("menu_icons/Rand.png"));
        randomBtn.addActionListener(e -> getWorkspaceComponent().getProjector().randomize(100));

        JToolBar theToolBar = new JToolBar();
        theToolBar.add(projectionList);
        playBtn.setToolTipText("Iterate projection algorithm");
        theToolBar.add(playBtn);
        iterateBtn.setToolTipText("Step projection algorithm");
        theToolBar.add(iterateBtn);
        clearBtn.setToolTipText("Clear current data");
        theToolBar.add(clearBtn);
        randomBtn.setToolTipText("Randomize datapoints");
        theToolBar.add(randomBtn);
        theToolBar.addSeparator();
        warningLabel.setPreferredSize(new Dimension(16, 16));
        warningLabel.setToolTipText("This method works best with more datapoints already added");
        theToolBar.add(warningLabel);
        String stepSizeToolTip = "Scales the amount points are moved on each iteration";
        JLabel stepSizeLabel = new JLabel("Step Size");
        stepSizeLabel.setToolTipText(stepSizeToolTip);

        sammonStepSizePanel.add(stepSizeLabel);
        JTextField sammonStepSize = new JFormattedTextField("" + SimbrainPreferences.getDouble(
                "projectorSammonEpsilon"));
        sammonStepSize.setColumns(3);
        sammonStepSize.setToolTipText(stepSizeToolTip);
        sammonStepSizePanel.add(sammonStepSize);
        theToolBar.add(sammonStepSizePanel);

        adjustDimension1.setToolTipText("Dimension 1");
        adjustDimension2.setToolTipText("Dimension 2");
        theToolBar.add(adjustDimension1);
        theToolBar.add(adjustDimension2);

        // Help button
        JButton helpButton = new JButton();
        helpButton.setAction(new ShowHelpAction("Pages/Plot/projection.html"));

        // Setup Menu Bar
        createAttachMenuBar();

        JToolBar statusBar = new JToolBar();
        statusBar.add(pointsLabel);
        statusBar.add(dimsLabel);

        JToolBar errorBar = new JToolBar();
        errorBar.add(errorLabel);

        JPanel southPanel = new JPanel();
        southPanel.add(errorBar);
        southPanel.add(statusBar);

        Box bottomPanel = Box.createVerticalBox();
        bottomPanel.add("South", southPanel);

        // Put all panels together
        add("North", theToolBar);
        add("Center", chartPanel);
        add("South", bottomPanel);

        // Other initialization
        initializeComboBoxes();

        proj.getEvents().onPointFound(p -> update());
        proj.getEvents().onDataChanged(() -> {
            resetData();
            update();
        });
        proj.getEvents().onColorsChanged(() -> {
            proj.resetColors();
            update();
        });

        // Epsilon field should update the model whenever a user clicks out of
        // it.
        sammonStepSize.addFocusListener(new FocusAdapter() {
            @Override
            public void focusLost(FocusEvent e) {
                ProjectionMethod projMethod = getWorkspaceComponent().getProjector().getProjectionMethod();
                if (projMethod != null) {
                    if (projMethod instanceof ProjectSammon) {
                        ((ProjectSammon) projMethod).setEpsilon(Utils.doubleParsable(sammonStepSize.getText()));
                    }
                }
            }
        });
        resetData();
        update();
        updateToolBar();
    }

    /**
     * Initialize all the combo boxes.
     */
    private void initializeComboBoxes() {

        // Populate projection list combo box
        for (Entry<Class<?>, String> projMethod : getWorkspaceComponent().getProjector().getProjectionMethods().entrySet()) {
            projectionList.addItem(projMethod.getValue());
        }
        projectionList.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                String selectedMethod = (String) projectionList.getSelectedItem();
                getWorkspaceComponent().getProjector().setProjectionMethod(selectedMethod);
                updateToolBar();
            }
        });
        projectionList.getModel().setSelectedItem(getWorkspaceComponent().getProjector().getCurrentMethodString());

        // Init the adjust dimension combo boxes
        updateCoordinateProjectionComboBoxes();
        adjustDimension1.setModel(adjustDimension1Model);
        adjustDimension1.addActionListener(e -> {
            ProjectionMethod projMethod = getWorkspaceComponent().getProjector().getProjectionMethod();
            if (projMethod != null) {
                if (projMethod instanceof ProjectCoordinate) {
                    ((ProjectCoordinate) projMethod).setHiD1(adjustDimension1.getSelectedIndex());
                    projMethod.project();
                    getWorkspaceComponent().getProjector().getEvents().fireDataChanged();
                }
            }
        });
        adjustDimension2.setModel(adjustDimension2Model);
        adjustDimension2.addActionListener(e -> {
            ProjectionMethod projecMethod = getWorkspaceComponent().getProjector().getProjectionMethod();
            if (projecMethod != null) {
                if (projecMethod instanceof ProjectCoordinate) {
                    ((ProjectCoordinate) projecMethod).setHiD2(adjustDimension2.getSelectedIndex());
                    projecMethod.project();
                    getWorkspaceComponent().getProjector().getEvents().fireDataChanged();
                }
            }
        });
    }

    /**
     * Update the toolbar based on current projection method.
     */
    private void updateToolBar() {
        ProjectionMethod proj = getWorkspaceComponent().getProjector().getProjectionMethod();
        if (proj == null) {
            return;
        }

        // Clear unused toolbar items
        if (!(proj instanceof ProjectSammon)) {
            sammonStepSizePanel.setVisible(false);
        }
        if (!(proj instanceof ProjectCoordinate)) {
            adjustDimension1.setVisible(false);
            adjustDimension2.setVisible(false);
        } else {
            // Set correct dimensions
            adjustDimension1.setSelectedIndex(((ProjectCoordinate) proj).getHiD1());
            adjustDimension2.setSelectedIndex(((ProjectCoordinate) proj).getHiD2());
        }

        // Handle error bar
        if ((proj.isIterable()) && (showError)) {
            errorBar.setVisible(true);
        } else {
            errorBar.setVisible(false);
        }

        // Handle warning
        if (getWorkspaceComponent().getProjector().getNumPoints() < proj.suggestedMinPoints()) {
            warningLabel.setVisible(true);
        } else {
            warningLabel.setVisible(false);
        }

        // Handle new toolbar items
        if (proj instanceof ProjectSammon) {
            sammonStepSizePanel.setVisible(true);
        } else if (proj instanceof ProjectCoordinate) {
            adjustDimension1.setVisible(true);
            adjustDimension2.setVisible(true);
        }

        // Handle iterable
        setToolbarIterable(proj.isIterable());

    }

    /**
     * Update the Coordinate projection combo boxes.
     */
    private void updateCoordinateProjectionComboBoxes() {

        adjustDimension1Model.removeAllElements();
        adjustDimension2Model.removeAllElements();
        int dims = getWorkspaceComponent().getProjector().getDimensions();
        for (int i = 0; i < dims; i++) {
            adjustDimension1Model.addElement(i + 1);
            adjustDimension2Model.addElement(i + 1);
        }
    }

    /**
     * Creates the menu bar.
     */
    private void createAttachMenuBar() {

        final JMenuBar bar = new JMenuBar();
        final JMenu fileMenu = new JMenu("File");

        PlotActionManager actionManager = new PlotActionManager(this);
        for (Action action : actionManager.getOpenSavePlotActions()) {
            fileMenu.add(action);
        }
        fileMenu.addSeparator();
        final JMenu exportImport = new JMenu("Export/Import...");
        fileMenu.add(exportImport);
        exportImport.add(ProjectionPlotActions.getImportData(getWorkspaceComponent().getProjector()));
        exportImport.addSeparator();
        exportImport.add(ProjectionPlotActions.getExportDataHi(getWorkspaceComponent().getProjector()));
        exportImport.add(ProjectionPlotActions.getExportDataLow(getWorkspaceComponent().getProjector()));
        fileMenu.addSeparator();
        fileMenu.add(new CloseAction(this.getWorkspaceComponent()));

        final JMenu editMenu = new JMenu("Edit");
        final JMenuItem preferencesGeneral = new JMenuItem("Preferences...");
        preferencesGeneral.addActionListener(e -> {
            ProjectionPreferencesDialog dialog = new ProjectionPreferencesDialog(getWorkspaceComponent().getProjector());
            dialog.pack();
            dialog.setLocationRelativeTo(null);
            dialog.setVisible(true);
        });
        editMenu.add(preferencesGeneral);


        final JMenuItem colorPrefs = new JMenuItem("Datapoint Coloring...");
        colorPrefs.addActionListener(e -> {
            DataPointColoringDialog dialog = new DataPointColoringDialog(getWorkspaceComponent().getProjector());
            dialog.pack();
            dialog.setLocationRelativeTo(null);
            dialog.setVisible(true);
        });
        editMenu.add(colorPrefs);

        final JMenuItem dims = new JMenuItem("Set dimensions...");
        dims.addActionListener(e -> {
            String dimensions = JOptionPane.showInputDialog("Dimensions:");
            if (dimensions != null) {
                getWorkspaceComponent().getProjector().init(Integer.parseInt(dimensions));
            }

        });
        // editMenu.add(dims);

        JMenu helpMenu = new JMenu("Help");
        JMenuItem helpItem = new JMenuItem(new ShowHelpAction("Pages/Plot/projection.html"));
        helpMenu.add(helpItem);

        bar.add(fileMenu);
        bar.add(editMenu);
        bar.add(helpMenu);

        getParentFrame().setJMenuBar(bar);
    }

    /**
     * Update labels at bottom of component.
     */
    protected void update() {
        chart.fireChartChanged();
        updateToolBar();
        dimsLabel.setText("     Dimensions: " + getWorkspaceComponent().getProjector().getUpstairs().getDimensions());
        pointsLabel.setText("  Datapoints: " + getWorkspaceComponent().getProjector().getDownstairs().getNumPoints());
        if (getWorkspaceComponent().getProjector().getProjectionMethod().isIterable()) {
            errorLabel.setText(" Error:" + ((IterableProjectionMethod) getWorkspaceComponent().getProjector().getProjectionMethod()).getError());
        }
        repaint();
    }

    /**
     * Enable or disable buttons depending on whether the current projection
     * algorithm allows for iterations or not.
     *
     * @param b whether the current projection algorithm can be iterated
     */
    private void setToolbarIterable(final boolean b) {
        if (b) {
            playBtn.setEnabled(true);
            iterateBtn.setEnabled(true);
        } else {
            playBtn.setEnabled(false);
            iterateBtn.setEnabled(false);
        }
    }

    /**
     * Custom rendering of scatter plot points
     */
    private class CustomRenderer extends XYLineAndShapeRenderer {

        @Override
        public Paint getItemPaint(int row, int column) {
            Projector projector = getWorkspaceComponent().getProjector();
            if (column >= projector.getNumPoints()) {
                System.out.println("getItemPaint:" + column + ">" + projector.getNumPoints());
                return Color.green;
            }
            DataPointColored point = ((DataPointColored) projector.getUpstairs().getPoint(column));
            if (point != null) {
                return point.getColor();
            } else {
                return Color.green;
            }
        }

    }

    /**
     * Custom label generator that renders a point's label, if any.
     */
    public class LegendXYItemLabelGenerator extends StandardXYItemLabelGenerator
            implements XYItemLabelGenerator {

        @Override
        public String generateLabel(XYDataset dataset, int series, int item) {
            Projector projector = getWorkspaceComponent().getProjector();
            if (item >= projector.getNumPoints()) {
                System.out.println("generateLabel:" + item + ">" + projector.getNumPoints());
                return null;
            }
            DataPointColored point = ((DataPointColored) projector.getUpstairs().getPoint(item));
            if (point != null) {
                return point.getLabel();
            }
            return null;
        }
    }

    /**
     * Datapoints return a tooltip showing the high dimensional point being
     * Represented by a given point in the plot.
     */
    private class CustomToolTipGenerator extends CustomXYToolTipGenerator {
        @Override
        public String generateToolTip(XYDataset data, int series, int item) {
            Projector projector = getWorkspaceComponent().getProjector();
            if (item >= projector.getNumPoints()) {
                System.out.println("generateToolTip:" + item + ">" + projector.getNumPoints());
                return null;
            }
            DataPoint point = projector.getUpstairs().getPoint(item);
            if (point != null) {
                return Utils.doubleArrayToString(point.getVector());
            } else {
                return "null";
            }
        }
    }

    public void resetData() {
        EventQueue.invokeLater(() -> {
            xyCollection.getSeries(0).clear();
            int size = getWorkspaceComponent().getProjector().getNumPoints();
            for (int i = 0; i < size; i++) {
                DataPoint point = getWorkspaceComponent().getProjector().getDownstairs().getPoint(i);
                xyCollection.getSeries(0).add(point.get(0), point.get(1));
            }
        });

    }

}

File: src/main/kotlin/org/simbrain/plot/projection/ProjectionDesktopComponent.kt
package org.simbrain.plot.projection

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.swing.Swing
import kotlinx.coroutines.withContext
import org.jfree.chart.ChartFactory
import org.jfree.chart.ChartPanel
import org.jfree.chart.JFreeChart
import org.jfree.chart.labels.CustomXYToolTipGenerator
import org.jfree.chart.labels.StandardXYItemLabelGenerator
import org.jfree.chart.labels.XYItemLabelGenerator
import org.jfree.chart.plot.PlotOrientation
import org.jfree.chart.renderer.xy.XYLineAndShapeRenderer
import org.jfree.data.xy.XYDataset
import org.jfree.data.xy.XYSeries
import org.jfree.data.xy.XYSeriesCollection
import org.simbrain.util.*
import org.simbrain.util.genericframe.GenericFrame
import org.simbrain.util.projection.*
import org.simbrain.util.widgets.ToggleButton
import org.simbrain.workspace.gui.DesktopComponent
import java.awt.*
import java.awt.geom.Ellipse2D
import javax.swing.*
import kotlin.reflect.full.primaryConstructor

class ProjectionDesktopComponent(frame: GenericFrame, component: ProjectionComponent)
    : DesktopComponent<ProjectionComponent>(frame, component), CoroutineScope {

    val projector = component.projector
    override var coroutineContext = projector.coroutineContext

    var running = false

    /**
     * Ordered list of [DataPoint] points so that the renderer can access points by index.
     */
    val pointList = ArrayList<DataPoint>()

    // Actions
    val iterateAction = createAction(
        iconPath = "menu_icons/Step.png",
        name = "Iterate",
        description = "Iterate once"
    ) {
        iterate()
    }

    val runAction = createAction(
        iconPath = "menu_icons/Play.png",
        name = "Run",
        description = "Run",
        coroutineScope = projector
    ) {
        if (!running) {
            running = true
            projector.events.startIterating.fire()
            launch {
                while (running) {
                    iterate()
                }
            }
        }
    }

    val stopAction = createAction(
        iconPath = "menu_icons/Stop.png",
        name = "Stop",
        description = "Stop"
    ) {
        running = false
        projector.events.stopIterating.fire()
    }

    val prefsAction = createAction(
        iconPath = "menu_icons/Prefs.png",
        name = "Preferences...",
        description = "Set projection preferences"
    ) {
        showPrefDialog()
    }

    val randomizeAction = createAction(
        name = "Randomize",
        description = "Randomize points",
        iconPath = "menu_icons/Rand.png"
    ) {
        projector.dataset.randomizeDownstairs()
        projector.events.datasetChanged.fireAndForget()
    }

    val clearDataAction = createAction(
        name = "Clear",
        description = "Clear all points",
        iconPath = "menu_icons/Eraser.png"
    ) {
        synchronized(projector.dataset) {
            projector.dataset.kdTree.clear()
            projector.events.datasetChanged.fireAndForget()
            projector.events.datasetCleared.fireAndForget()
        }
    }

    // Top stuff
    val projectionMethods = projectionTypes
        .associateWith { it.kotlin.primaryConstructor!!.call() }
    val projectionSelector = JComboBox<ProjectionMethod>().apply {
        maximumSize = Dimension(200, 100)
        projectionMethods.values.forEach {
            addItem(it)
        }.also {
            addActionListener {
                projector.projectionMethod = (selectedItem as ProjectionMethod)
            }
        }
        selectedItem = projectionMethods[projector.projectionMethod.javaClass]
    }
    val mainToolbar = JToolBar().apply {
        add(projectionSelector)
        addSeparator()
        add(prefsAction)
        add(randomizeAction)
        add(clearDataAction)
    }
    private val runToolbar = JToolBar().apply {
        add(ToggleButton(listOf(stopAction, runAction)).apply {
            setAction("Run")
            projector.events.startIterating.on {
                setAction("Stop")
            }
            projector.events.stopIterating.on {
                setAction("Run")
            }
        })
        add(iterateAction)
    }

    val topPanel = JPanel(FlowLayout(FlowLayout.LEFT)).also {
        it.add(mainToolbar)
    }

    // Central Chart Panel
    private suspend fun iterate() {
        projector.projectionMethod.let { projection ->
            if (projection is IterableProjectionMethod) {
                projection.iterate(projector.dataset)
                projector.events.iterated.fire(projection.error)
            }
        }
        projector.events.datasetChanged.fire()
    }

    /**
     * JChart representation of the data.
     */
    private val xyCollection: XYSeriesCollection = XYSeriesCollection().apply {
        addSeries(XYSeries("Data", false, true))
    }

    private val renderer = CustomRenderer(this).apply {
        setSeriesLinesVisible(0, projector.connectPoints)
        setSeriesShape(0, Ellipse2D.Double(-7.0, -7.0, 7.0, 7.0))
        val generator = CustomToolTipGenerator(this@ProjectionDesktopComponent)
        setSeriesToolTipGenerator(0, generator)
        defaultItemLabelsVisible = true
        defaultItemLabelGenerator = LegendXYItemLabelGenerator(this@ProjectionDesktopComponent)
    }


    /**
     * The JFreeChart chart.
     */
    private val chart: JFreeChart = ChartFactory.createScatterPlot(
        "", "Projection X", "Projection Y",
        xyCollection, PlotOrientation.VERTICAL, false, true, false
    ).apply {
        xyPlot.backgroundPaint = Color.white
        xyPlot.domainGridlinePaint = Color.gray
        xyPlot.rangeGridlinePaint = Color.gray
        xyPlot.domainAxis.isAutoRange = true
        xyPlot.rangeAxis.isAutoRange = true
        xyPlot.foregroundAlpha = .5f // TODO: Make this settable
        xyPlot.renderer = renderer
    }
    val chartPanel = ChartPanel(chart).also {
        add(it)
    }

    // Bottom stuff
    val pointsLabel = JLabel()
    val dimensionsLabel = JLabel()
    val errorLabel = JLabel("Error: ---")
    val bottomPanel = JPanel().apply {
        layout = FlowLayout(FlowLayout.LEFT)
        add(pointsLabel)
        add(Box.createHorizontalStrut(25));
        add(dimensionsLabel)
        add(Box.createHorizontalStrut(25));
    }

    fun showPrefDialog() {
        projector.createEditorDialog {
            it.init()
            launch {
                it.events.settingsChanged.fire()
                redrawAllPoints()
            }
        }.display()
    }

    private suspend fun redrawAllPoints() {
        withContext(Dispatchers.Swing) {
            xyCollection.getSeries(0).clear()
            pointList.clear()
            projector.dataset.kdTree.forEach {
                pointList.add(it)
                val (x, y) = it.downstairsPoint
                xyCollection.getSeries(0).add(x, y)
            }
            pointsLabel.text = "Datapoints: ${projector.dataset.kdTree.size}"
            dimensionsLabel.text = "Dimensions: ${projector.dimension}"
        }
    }

    init {
        layout = BorderLayout()

        add("North", topPanel)
        add("Center", chartPanel)
        add("South", bottomPanel)

        frame.jMenuBar = JMenuBar().apply {
            add(JMenu("File").apply {
                add(JMenuItem(importAction))
                add(JMenuItem(exportAction))
                addSeparator()
                add(JMenuItem(closeAction))
            })
            add(JMenu("Edit").apply {
                val prefsAction: Action = prefsAction
                add(JMenuItem(prefsAction))
            })
        }

        projector.events.datasetChanged.on {
            redrawAllPoints()
        }
        projector.events.pointUpdated.on {
            chart.fireChartChanged()
        }
        projector.events.datasetCleared.on {
            projector.coloringManager.reset()
        }
        projector.events.settingsChanged.on {
            renderer.setSeriesLinesVisible(0, projector.connectPoints)
        }
        projector.events.methodChanged.on { o, n ->
            projectionSelector.selectedItem = projectionMethods[n.javaClass]
            running = false
            projector.events.stopIterating.fireAndBlock()
            if (n is IterableProjectionMethod) {
                topPanel.add(runToolbar)
                bottomPanel.add(errorLabel)
            } else {
                bottomPanel.remove(errorLabel)
                topPanel.remove(runToolbar)
            }
            topPanel.revalidate()
            topPanel.repaint()
            bottomPanel.revalidate()
            bottomPanel.repaint()
            launch { redrawAllPoints() }
        }
        projector.events.iterated.on { error ->
            errorLabel.text = "Error: ${error.format(2)}"
        }
        launch {
            redrawAllPoints()
        }
    }

}

fun main() {
    val projector = Projector(5).apply {
        // val random = Random(1)
        // repeat(100) {
        //     projector.addDataPoint(DoubleArray(5) { random.nextDouble() })
        // }
        // projector.projectionMethod = SammonProjection2(projector.dimension).apply {
        //     epsilon = 100.0
        // }
        (0 until 40).forEach {p ->
            val point= DataPoint(DoubleArray(100) { p.toDouble() }, label = "$p")
            addDataPoint(point)
        }
        dataset.randomizeDownstairs()
        init()
    }
    StandardDialog().apply{
        val desktopComponent = ProjectionDesktopComponent(
            this, ProjectionComponent("test", projector))
        contentPane = desktopComponent
        makeVisible()
    }
}

private class CustomRenderer(val proj: ProjectionDesktopComponent) : XYLineAndShapeRenderer() {
    override fun getItemPaint(series: Int, index: Int): Paint {
        val projector = proj.projector
        val hotColor = if (projector.useHotColor) projector.hotColor else projector.baseColor
        if (proj.pointList[index] === proj.projector.dataset.currentPoint) {
            return hotColor
        }
        return with(projector) {
            projector.coloringManager.getColor(proj.pointList[index])?: projector.baseColor
        }
    }
}

private class CustomToolTipGenerator(val proj: ProjectionDesktopComponent) : CustomXYToolTipGenerator() {
    override fun generateToolTip(data: XYDataset, series: Int, index: Int): String {
        return proj.pointList[index].upstairsPoint.format(2)
    }
}

class LegendXYItemLabelGenerator(val proj: ProjectionDesktopComponent) : StandardXYItemLabelGenerator(), XYItemLabelGenerator {
    override fun generateLabel(dataset: XYDataset, series: Int, index: Int): String? {
        return if (proj.projector.showLabels) proj.pointList[index].label else null
    }
}

